### PR TITLE
Add support for SQL bitwise operators on integer types

### PR DESCRIPTION
**Pull request by aalexandrov**
### Motivation

Adds missing bitwise operators on integers as suggested in #7192.

### Description

Adds the following bitwise operators on `smallint`, `int`, and `bigint` types

- [x] Bitwise AND (`&`)
- [x] Bitwise XOR (`#`)
- [x] Bitwise OR (`|`)
- [x] Bitwise NOT (`~`)
- [x] Shift left (`<<`)
- [x] Shift right (`>>`)

### Tips for reviewer

- I used #5966 to figure out what changes are needed. 
- I had to add a new precedence for the bitwise NOT in order to maintain compatibility with Postgres (see diff in `parser.rs`). Note that this is different [from the CockroachDB semantics](https://www.cockroachlabs.com/docs/stable/postgresql-compatibility.html#precedence-of-unary) which are not aligned with Postgres.

### Checklist

- [x] This PR has adequate test coverage.
- [x] This PR adds a release note for any user-facing behavior changes.

### CODE REVIEW

File: doc/developer/overview.md
Comment (philip-stoev) [inline doc/developer/overview.md:110]: Word truncated here . Do those changes belong to this PR?

File: test/sqllogictest/arithmetic.slt
Comment (philip-stoev) [inline test/sqllogictest/arithmetic.slt:569]: Thank you for the tests! You can also display/make explicit the precedence using an EXPLAIN statement of a SELECT that uses actual columns (so the expression are not constant-folded). The EXPLAIN should show the order of operations using nested brackets on the `Map` line:
```
materialize=> explain select f1 + f1 - f1 from t1;
           Optimized Plan            
-------------------------------------
 %0 =                               +
 | Get materialize.public.t1 (u1667)+
 | Map ((#0 + #0) - #0)             +
 | Project (#1)                     +
```

**Review by philip-stoev — APPROVED** [review]

File: src/sql/src/func.rs
Comment (aljoscha) [inline src/sql/src/func.rs:2350]: I think for `smallint` it should be `1874`

File: doc/developer/overview.md
Comment (aalexandrov) [reply doc/developer/overview.md:110]: These are doc changes that I did as part of the onboarding. I will move them in a separate PR.

File: test/sqllogictest/arithmetic.slt
Comment (aalexandrov) [reply test/sqllogictest/arithmetic.slt:569]: Didn't think of that, good point!

File: src/sql/src/func.rs
Comment (aalexandrov) [reply src/sql/src/func.rs:2350]: Good catch.

File: src/expr/src/scalar/func.rs
Comment (antiguru) [inline src/expr/src/scalar/func.rs:1169]: Could the unary funcs be expressed as a [macro](https://github.com/MaterializeInc/materialize/blob/8e8b351490b2bf4cd5e4ecccfa37953c6dee5da4/src/expr/src/scalar/func/macros.rs#L10)? Not sure tbh, but maybe worth checking!

Comment (aalexandrov) [reply src/expr/src/scalar/func.rs:1169]: Probably, but this issue is more general, as most functions come in families indexed by the possible combination of types (e.g. see `neg_int{XX}`, `abs_int{XX}`, ...). I agree that we should think how we can minimize code for this functions, but I prefer to not tackle this as part of my starter issue. We could use a macro, think about how to utilize generics in a better way, or choose another zero cost abstraction.

I'm happy to open a follow-up issue if more people agree that this is a problem.

Comment (philip-stoev) [reply src/expr/src/scalar/func.rs:1169]: @petrosagg ^^^

Comment (antiguru) [reply src/expr/src/scalar/func.rs:1169]: I was thinking about something like this:
```
sqlfunc!(
    fn bit_not_int16(a: i16) -> i16 {
        !a
    }
);
```

Elsewhere, there seems to be a `TryFrom` missing:
```
impl TryFrom<Datum<'_>> for i16 {
    type Error = ();
    fn try_from(from: Datum<'_>) -> Result<Self, Self::Error> {
        match from {
            Datum::Int16(f) => Ok(f),
            _ => Err(()),
        }
    }
}

impl TryFrom<Datum<'_>> for Option<i16> {
    type Error = ();
    fn try_from(from: Datum<'_>) -> Result<Self, Self::Error> {
        match from {
            Datum::Null => Ok(None),
            Datum::Int16(f) => Ok(Some(f)),
            _ => Err(()),
        }
    }
}
```

But, totally up to you :D

Comment (petrosagg) [reply src/expr/src/scalar/func.rs:1169]: Using the `sqlfunc!` macro for this unary function is totally doable and totally optional since this is your first issue. We're in the process of transitioning the traditional declarations (like the one you provided for `bit_not_int16`) to use `sqlfunc!`. The plan is to transition all functions (even binary and variadic), but for now only unarfy functions are supported by the macro.

So, if you want to merge as is, that's totally fine, I will include this function in a future PR that will migrate all the `int16` functions whole sale.

If you want to dip your toes in the new way of defining unary functions you can use @antiguru's snippets above and a commit like this one for reference: https://github.com/MaterializeInc/materialize/pull/8271/commits/88a144309dc71cca8c350e18570bd9f7fdc6f2b3

File: src/sql-parser/src/parser.rs
Comment (petrosagg) [inline src/sql-parser/src/parser.rs:155]: Is the precedence different than `Other`?

Comment (aalexandrov) [reply src/sql-parser/src/parser.rs:155]: Actually according to [this table](https://www.postgresql.org/docs/7.2/sql-precedence.html) in postgres all bitwise operators should actually have precedence `Other`, so I think that you are right and we can remove this.

Comment (aljoscha) [reply src/sql-parser/src/parser.rs:155]: Does that mean the comment in the Cockroach docs about prededence of `~` was wrong?

Comment (aalexandrov) [reply src/sql-parser/src/parser.rs:155]: This is from [the Cockroach compatibility page](https://www.cockroachlabs.com/docs/stable/postgresql-compatibility.html#precedence-of-unary):

> In PostgreSQL, the unary ~ (bitwise not) operator has a low precedence. For example, the following query is parsed as ~ (1 + 2) because ~ has a lower precedence than +:

This statement is line with [the table referenced above](https://www.postgresql.org/docs/7.2/sql-precedence.html) (as well as our precedence table) where only the binary `+` and `-` have higher precedence than `~`, but `~` has the same precedence as `&` and `|` (everything falls under `other`). 

Based on that information as expected the following equations hold in Postgres:

```
~x | y = (~x) | y
~x - y = ~(x - y)
``` 

The are tests marked with the headers

```
# precedence between '~' and ("|", "&")
```
and
```
# precedence between '~' and ('+', '-')
```

ensure that this is also the case in Materialize.

File: src/expr/src/scalar/func.rs
Comment (aalexandrov) [reply src/expr/src/scalar/func.rs:1169]: > So, if you want to merge as is, that's totally fine,  I will include this function in a future PR that will migrate all the int16 functions whole sale.

@petrosagg: I like the approach outlined in #7704! Unfortunately, missed that and started working on a this feature shortly before the PR was merged, so I prefer to merge the current PR without further. I am happy take off some of the refactoring work (including the `~` operator) in the next days.

> Elsewhere, there seems to be a TryFrom missing:

@antiguru Is the above somehow affecting the correctness of the current implementation and actionable on my end?

Comment (petrosagg) [reply src/expr/src/scalar/func.rs:1169]: No, `TryFrom` is only needed if you go down the `sqlfunc!` path. Don't worry, I'm slowly converting all the functions in batches, let's merge this as is :)

File: doc/user/content/release-notes.md
Comment (petrosagg) [inline doc/user/content/release-notes.md:52]: extra new line here

File: src/expr/src/scalar/func.rs
Comment (petrosagg) [inline src/expr/src/scalar/func.rs:1224]: Rust will panic if `b` is negative but postgres has overflow semantics. You can use [`overflow_shl`](https://doc.rust-lang.org/std/primitive.i32.html#method.overflowing_shl) and friends in all these functions to have the same behaviour.

Also, it took me a while to figure out what's going on with the `& 0b_0001_1111` part and from some quick tests it doesn't seem necessary to me. This produces the same results as postgres for me:

```rust
    a.unwrap_i32().overflowing_shl(b.unwrap_i32() as u32).0
```

File: doc/user/content/sql/functions/_index.md
Comment (petrosagg) [inline doc/user/content/sql/functions/_index.md:60]: Missing shift operators

File: src/expr/src/scalar/func.rs
Comment (petrosagg) [inline src/expr/src/scalar/func.rs:1219]: Is there a link to the pg docs about this that can be included in this comment?

File: src/expr/src/scalar/func.rs
Comment (petrosagg) [reply src/expr/src/scalar/func.rs:1224]: I noticed today that the bitmask also handles negatives so that's not an issue with the current code but I still think we should use `overflowing_shl` since that's the behaviour we're after

**Review by antiguru — APPROVED** [review]
LGTM once remaining comments are addressed! Thanks!

File: doc/user/content/release-notes.md
Comment (aalexandrov) [reply doc/user/content/release-notes.md:52]: I followed the formatting from the last few versions (line items are separated by a new line). I now realize this is not correct markdown, though (it will be interpreted as a sequence of singleton lists, so maybe we should omit the newlines as you are suggesting.

File: src/expr/src/scalar/func.rs
Comment (aalexandrov) [reply src/expr/src/scalar/func.rs:1219]: Sadly this part is based on empirical analysis on the behavior of Postgres in cases where the number of shifted bits exceeds the integer length. I've added tests in `arithmetic.slt` ([here](https://github.com/aalexandrov/materialize/blob/sql_bitwise_operators/test/sqllogictest/arithmetic.slt#L589) and [here](https://github.com/aalexandrov/materialize/blob/sql_bitwise_operators/test/sqllogictest/arithmetic.slt#L589)).

I can reference the tests file or inline a specific example.

File: src/expr/src/scalar/func.rs
Comment (aalexandrov) [reply src/expr/src/scalar/func.rs:1224]: > Rust will panic if b is negative but postgres has overflow semantics

I am covering negative `b` values in the tsts ([here](https://github.com/aalexandrov/materialize/blob/sql_bitwise_operators/test/sqllogictest/arithmetic.slt#L589) and [here](https://github.com/aalexandrov/materialize/blob/sql_bitwise_operators/test/sqllogictest/arithmetic.slt#L589)) and they work as expected (result tables are taken from Postgres).

Will take a look at the `overflowing_shl` & friends.

File: src/expr/src/scalar/func.rs
Comment (petrosagg) [reply src/expr/src/scalar/func.rs:1219]: Hm, you can link to this section of pg source code https://github.com/postgres/postgres/blob/REL_14_STABLE/src/backend/utils/adt/int.c#L1460-L1476

Interestingly [`i16::overflowing_shl`](https://doc.rust-lang.org/std/primitive.i16.html#method.overflowing_shl) takes a u32 argument so this can also be written like so:

```rust
a.unwrap_int16().oveflowing_shl(b.unwrap_i32() as u32).0
```

Comment (aalexandrov) [reply src/expr/src/scalar/func.rs:1219]: According to Postgres we should get:

```
1::int2 <<  1 = 2
1::int2 << 15 = -32768
1::int2 << 16 = 0
1::int2 << 17 = 0 
...
1::int2 << 31 = 0
1::int2 << 32 = 1
1::int2 << 33 = 1
```
and the way to get this behavior is apply the `<<` operation on a lhs that is temporarily widened from `i16` to `i32`.

**Review by petrosagg — APPROVED** [review]

File: src/expr/src/scalar/func.rs
Comment (aalexandrov) [reply src/expr/src/scalar/func.rs]: After further discussion and digging by @petrosagg, seems like an instance of [C's integer promotion rules](http://www.idryman.org/blog/2012/11/21/integer-promotion/#:~:text=In%20C99%2C%20integer%20promotion%20is,are%20called%20the%20integer%20promotions.) in action.

Put a comment on that in the `bit_shift_left_int16` and `bit_shift_right_int16` bodies.

**Pull request merged**
